### ISSUE TITLE
Adaptive Learning Rate

### ISSUE THREAD

**Issue by rladeira** [dmlc/xgboost#1042]
In the context of xgboost, does adaptive learning rate (eta parameter) make sense?

I was wondering if some algorithm for this could help, since sometimes we need eta very small to obtain optimal results, so that, a lot of computational effort can be wasted in the first rounds, without progressing a lot towards the final solution.

**Comment by tqchen** [comment]
I have no experience in this, so won't make comment. But if anyone have experience trying the setting is welcomed to share theirs

**Comment by thenomemac** [comment]
Agreed this would be useful and essentially eliminate number of trees as a parameter. I suggest something similar to learning rate annealing, widely used in neural networks. For example, start the learning rate at .2 or .1 then every time a step leads to over fitting either on a validation set or on OOB samples you could half the learning rate (could be any number maybe halving the learning rate is too aggressive especially if using OOB samples). This way the learning rate would quickly converge to zero as the trees start to over fit. As a result you could theoretically get close to the same performance with many fewer trees. This strategy is similar to the stacking of GBMs that's become popular on kaggle.

**Comment by rladeira** [comment]
@thenomemac, That's the intuition behind the idea: doing something similar to what is done in neural networks algorithms. As soon as I get the time, I'll try to code something implemeting the idea.

However, why are you considering this similar to stacking of multiple GBMS? Stacking is about using knowledge of multiples generalizers (GBMs, in this case) to build a better one. The adaptive learning rate aims to achieve better convergence faster, can't see why this would related to stacking.

**Comment by thenomemac** [comment]
I was trying to come up with additional justification for why this would work well in case core members object to adding this feature. I have seen people do something on kaggle where they run a GBM with learning rate of .2 and do like 100 tree, then take those predictions and feed them into a 100 iterations at .05 then feed those into 100 iterations at .01 ect. The idea being you need less total iterations to get the same model since in practice a low learning rate will initially split on the same features many times in a row.  

I think this feature will be super useful in reducing the overall clock time in ML pipelines. For example now I may need to run 5000 trees with .01 learning rate. This algorithm should in theory give me similar results with many fewer iterations. Kudos to you for coding something on this.

**Comment by thenomemac** [comment]
@rladeira I would also suggest that we allow this meathod to integrate with early stopping. I think it should natively. For example your leaning rate starts off at .5 by iteration 1000 it's at .001 well likely the overall fit will not improve much once the learning rate is that low so if you have your early stopping threshold at 50 iterations the model should just stop at what is basically the optimal iteration number.

**Comment by rladeira** [comment]
@thenomemac, I got it, you are talking about stacking in a higher level. I was thinking about stacked generalization, a differente idea, which is also used in Kaggle.

The example you described covers exactly the intuition behind the idea. Nice to know that it has been successfully applied in kaggle.

Yes, the early stopping should help a lot in this setting as well.

**Comment by rladeira** [comment]
@tqchen , I have been trying to modify xgb.cv to implement the idea discussed in this thread. Firstly, I'm trying to control the eta parameter based on test performance for cross validation. As soon as overfitting is detected, the learning rate is decreased. 

At some point, I need to store the best models found for each cv fold. When overfitting is detected, I need the best models saved in independents objects, so I can use them later. In this context, a copy of a model pointed by a "xgb.Booster.handle" object is needed. Is there a way i can accomplish this?

Another option would be to train whole new models for each cv partition, based on the best round computed. Although this is not the best option, since there's a lot extra computation, I think I'll go this way, if there is not other option.

**Comment by Far0n** [comment]
At least for python, it's already there: https://github.com/Far0n/xgboost/commit/a1ba6086417d58a735e638496a1932bacbf18651
